ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala

```diff
@@ -76,13 +80,51 @@ class ApiPackageManagementService(
             case r @ UploadPackagesResult.NotSupported =>
               Future.failed(ErrorFactories.unimplemented(r.description))
           }(DE)
+          .flatMap(
+            pollUntilPersisted(
+              res._2,
+              _,
+              50.milliseconds,
+              500.milliseconds,
+              (d: FiniteDuration) => d * 2))(DE)
     )
   }
+
+  /**
+    * Wraps a call [[PollingUtils.pollUntilPersisted]] so that it can be chained on the package upload with a `flatMap`.
+    *
+    * Checks invariants and forwards the original result after all packages are found to be persisted.
+    *
+    * @param ids The IDs of the uploaded packages
+    * @return The result of the party allocation received originally, wrapped in a [[Future]]
+    */
+  private def pollUntilPersisted(
+      ids: List[String],
+      result: UploadDarFileResponse,
+      minWait: FiniteDuration,
+      maxWait: FiniteDuration,
+      iteration: FiniteDuration => FiniteDuration): Future[UploadDarFileResponse] = {
+    val newPackages = ids.toSet
+    val description = s"packages ${ids.mkString(", ")}"
+    PollingUtils
+      .pollUntilPersisted(() => packagesIndex.listLfPackages())(
```

Comment:
```suggestion
      .pollUntilPersisted(packagesIndex.listLfPackages _)(
```
I guess it's a matter of taste, use your discretion.